specs/CONFIG.md

```diff
@@ -0,0 +1,199 @@
+# libp2p Daemon Config Spec
+
+The libp2p daemon is an interface to libp2p, and as such has a large number of 
+configuration options. For complex deployments, these options can become 
+unwieldy on the command line. For this, we allow the ability to provide a
+configuration spec via json from a file or from stdin
+
+_At the moment, this is a living document. As such, it will be susceptible to
+changes until stabilization._
+
+## Considerations
+
+As this is a spec that must be shared between multiple implementations, it is 
+probably best that the structure be generated from some shared source of truth,
+instead of depending on discipline to keep the implementation consistent 
+between language implementations. For instance, the config structure could be 
+generated as a protobuf message.
+
+## Command Line
+
+There are two ways to provide a JSON configuration to the daemon. Both methods 
+cause all other configuration command line options to be ignored.
+
+### Read from a file
+`p2pd -f ./conf.json`
+
+### Read from stdin
+`cat ./conf.json | p2pd -i`
+
+
+## Schema
+
+* `Field Name`
+    * Description
+    * Type
+    * `default`
+
+* `ListenAddr`
+    * Daemon control listen multiaddr
+    * Maddr String
+    * `"/unix/tmp/p2pd.sock"`
+* `Quiet`
+    * Be Quiet
+    * Boolean
+    * `false`
+* `ID`
+    * Peer identity; private key file
+    * String
+    * `""`
+* `Bootstrap`
+    * `Enabled`
+        * Connects to bootstrap peers and bootstraps the dht if enabled
+        * Boolean
+        * `false`
+    * `Peers`
+        * List of bootstrap peers; defaults to the IPFS DHT peers
+        * Array[Maddr String]
+        * `[]`
+* `DHT`
```

Comment:
As discussed earlier, we should be able to have more control on the DHT configuration. 

For example, we have the interop tests for the DHT content fetching blocked by not being able to add custom validators or selectors. Moreover, I think we should be able to add other options to the DHT, such as  the random walk config [libp2p/js-libp2p-kad-dht/src/index.js#L39-L42](https://github.com/libp2p/js-libp2p-kad-dht/blob/master/src/index.js#L39-L42) and the constants [libp2p/js-libp2p-kad-dhtmaster/src/constants.js#L32-L39](https://github.com/libp2p/js-libp2p-kad-dht/blob/master/src/constants.js#L32-L39). The constants being configurable is a topic that was already discussed a few times in the context of experiments with the testbed.

Reply:
That sounds good, but it also sounds like it's out of scope of this particular ticket. This config should be for the daemon as it currently stands. Any feature additions to the deamon will also need to include appropriate config edits

Reply:
All right, that sounds correct for me! Just mentioned to not fall out of the radar! 

It is also important to consider that as we are creating the config now, I think we should at least prepare the dht options to get the remaining configs. Otherwise, if we will expect a string for now, we will create a breaking change afterward.

Reply:
I edited the DHT property to be an object, and added a field `Mode` that is an enum of "client", "full" and empty string. It defaults to empty string. It should be easier to add extra fields now.